### PR TITLE
Secure group recommendations

### PR DESCRIPTION
**Pull request by jilljenn**
- [x] Speed up reco
- [x] ElGamal on embeddings

### CODE REVIEW

File: mangaki/mangaki/utils/recommendations.py [hunk 1]
```diff
@@ -159,30 +106,38 @@ def get_group_reco_algo(request, users_id=None, algo_name='als',
...
+        embeddings.append(transform(user_id,
+            algo.fit_single_user(
+                user_ratings['encoded_work_id'],
+                user_ratings['rating']
+            )))
+
+    if is_encrypted:
+        sum_means, sum_feats = he.decrypt_embeddings(embeddings)
+        group_parameters = [[(my_mean + sum_means) / group_length,
+                             (my_feat + sum_feats) / group_length]]
```
Comment (Akulen) [inline mangaki/mangaki/utils/recommendations.py:135]: This should probably be a tuple and not an array for consistency with the unencrypted case
```python
        group_parameters = [((my_mean + sum_means) / group_length,
                             (my_feat + sum_feats) / group_length)]
```

File: mangaki/mangaki/utils/crypto.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    :complexity: O(log a + log p)
+    """
+    return bezout(a, p)[0] % p
+
+
+class HomomorphicEncryption:
+    def __init__(self, user_ids, quantize_round=0, MAX_VALUE=10):
+        self.user_ids = user_ids
+        self.quantize_round = quantize_round
+        self.g = randint(1, PRIME - 1)
```
[the remark this answers is not in the document]
Comment (jilljenn) [reply mangaki/mangaki/utils/crypto.py:51]: 😂 Review de haut niveau

File: mangaki/mangaki/utils/crypto.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        self._keygen()
+        self._shares = {}
+        self._encode(MAX_VALUE)
+
+    def _keygen(self):
+        self._sk = {user: randint(0, PRIME) for user in self.user_ids}
```
[the remark this answers is not in the document]
Comment (jilljenn) [reply mangaki/mangaki/utils/crypto.py:57]: > 1 is also better

File: mangaki/mangaki/utils/crypto.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            self._shares[user_id][dim] = expmod(c1, self._sk[user_id])
+            encrypted.append(c2)
+        return np.array(encrypted)
+
+    def combine_embeddings(self, encrypted_embeddings):
+        combined = np.ones_like(encrypted_embeddings[0])
+        for encrypted_embedding in encrypted_embeddings:
+            combined *= encrypted_embedding
+            combined %= PRIME
+        self._combined_shares = np.ones_like(encrypted_embeddings[0], dtype=float)
```
[the remark this answers is not in the document]
Comment (jilljenn) [reply mangaki/mangaki/utils/crypto.py:91]: No indeed, I removed, I can't believe I didn't get an error

File: mangaki/mangaki/utils/recommendations.py [hunk 5]
```diff
@@ -9,20 +9,21 @@
...
 
-def get_algo_backup_or_fit_knn(algo_name):
+def get_algo_backup_or_fit_svd(algo_name):
     try:
         algo = get_algo_backup(algo_name)
     except FileNotFoundError:
+        # Fallback to SVD
         triplets = list(
             Rating.objects.values_list('user_id', 'work_id', 'choice'))
-        # In the future, we should warn the user it's gonna take a while
```
[the remark this answers is not in the document]
Comment (jilljenn) [reply mangaki/mangaki/utils/recommendations.py:23]: I tried something

File: mangaki/mangaki/utils/recommendations.py [hunk 6]
```diff
@@ -15,15 +17,22 @@
 CHRONO_ENABLED = True
 
 
-def get_algo_backup_or_fit_svd(algo_name):
+def get_algo_backup_or_fit_svd(request, algo_name):
```
Comment (Akulen) [inline mangaki/mangaki/utils/recommendations.py:20]: Do we need request here?

**Review by Akulen — APPROVED** [review]

**Pull request merged**